### ISSUE TITLE
CdiPoolPut() call leads to assert in CdiListRemove() when trying to remove an entry from an empty list.

### ISSUE THREAD

**Issue by yanniel** [aws/aws-cdi-sdk#31]
I do not have a deterministic way to reproduce it. It happens sporadically, but it is not extremely hard to reproduce (I get it a few times a day).

The use case is one Raw TX and one Raw RX in the same EC2 but on different processes. EFA adapter type and the loopback address are used.

Call stack attached.

![image](https://user-images.githubusercontent.com/55061834/106015950-fc544080-608c-11eb-9f51-f41e28eac6d2.png)

For reference: a situation leading to the same assert was recently fixed in commit: https://github.com/aws/aws-cdi-sdk/commit/26a06598c104159e4bae67365777bc818cb8e925.

OS: Windows Server 2019 Datacenter 64-bit (10.0 Build 17763)

Thanks.

**Comment by mhhen** [comment]
Since the pool is being used in application code, we don't have much visibility here. Typically, this problem occurs when a pool entry has been Put() more than once after only a single Get(). Are you able to provide any additional details? Is the problem in the Tx or Rx application?

**Comment by yanniel** [comment]
We haven't seen this in a while. Since last we saw it our application code changed significantly. Either these changes made this harder to reproduce or it was a bug in the app that was fixed. My gut feel in hindsight is that it was/is our bug; not an SDK issue. See "context" below for a theory of what could have happened.

We'll be running some tests to see if we can reproduce it once more. **We'll report back on this.**

Context: The issue was happening on the TX side. This is the only place where we deal with memory pools (for the purpose managing transmit buffers). The error came from the TX callback that's triggered when a payload is received by the RX. In this callback we were returning the transmit buffer that was just received by the RX to the pool. It's possible we were shutting down our CDI processing when we got this callback. Our shutting down procedure would run our clean-up code (that returns everything to the pool and destroys the pools), but the callback was triggered after the clean-up code. In which case we would have issued two Put() against a single Get(); leading to the assert.
